cpp/src/buffer/buffer.cpp

```diff
@@ -139,8 +155,135 @@ std::unique_ptr<Buffer> Buffer::copy(MemoryType target, rmm::cuda_stream_view st
     );
 }
 
+template <cudaMemcpyKind Kind, typename T>
+std::unique_ptr<Buffer> cuda_copy_slice_impl(
+    const T& storage,
+    std::ptrdiff_t offset,
+    std::size_t length,
+    MemoryReservation& target_reserv,
+    rmm::cuda_stream_view stream
+) {
+    // allocate buffer using the target reservation
+    auto out_buf = target_reserv.br()->allocate(
+        target_reserv.mem_type(), length, stream, target_reserv
+    );
+
+    if (length > 0) {
+        RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
+            static_cast<cuda::std::uint8_t*>(out_buf->data()),
+            static_cast<cuda::std::uint8_t*>(storage->data()) + offset,
+            length,
+            Kind,
+            stream
+        ));
+        // override the event to track the async copy, if the copy is not host-to-host
+        if constexpr (Kind != cudaMemcpyHostToHost) {
+            out_buf->override_event(std::make_shared<Buffer::Event>(stream));
+        }
```

Comment:
*question*: If the copy is host-to-host why do we not need an event? I thought that `cudaMemcpyAsync` is "asynchronous with respect to the host"?

Answer: https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html

Maybe reference that here.